File: apps/desktop/frontend/src/app/clipboard-history/models/clipboard-item.ts
import { signal } from '@angular/core';
import { dto } from '@wails/models';

export type ClipboardItemType = 'TEXT' | 'IMAGE' | 'URL';

export class ClipboardItem {
  readonly id: string;
  readonly type: ClipboardItemType;
  readonly content: string;
  readonly isPinned = signal(false);
  readonly createdAt: number;

  pinnedAt = 0;

  constructor(source: dto.ClipboardItem) {
    this.id = source.id;
    this.type = source.type as unknown as ClipboardItemType;
    this.content = source.content;
    this.isPinned.set(source.isPinned);
    this.createdAt = source.createdAt;

    if (this.isPinned()) {
      this.pinnedAt = Date.now();
    }
  }

  togglePin() {
    this.isPinned.set(!this.isPinned());

    this.pinnedAt = Date.now();
  }
}

File: apps/desktop/internal/clipboard/dto/clipboard_item.go
package dto

import (
	"encoding/json"
	"errors"
	"fmt"
)

type ClipboardItemType byte

const (
	ClipboardItemTypeUnknown ClipboardItemType = iota
	ClipboardItemTypeText
	ClipboardItemTypeImage
	ClipboardItemTypeUrl
)

func (itemType ClipboardItemType) MarshalJSON() ([]byte, error) {
	return []byte(`"` + itemType.String() + `"`), nil
}

func (itemType ClipboardItemType) String() string {
	switch itemType {
	case ClipboardItemTypeText:
		return "TEXT"
	case ClipboardItemTypeImage:
		return "IMAGE"
	case ClipboardItemTypeUrl:
		return "URL"
	case ClipboardItemTypeUnknown:
		return "UNKNOWN"
	}

	panic(fmt.Sprintf("unknown clipboard content item type '%d'", itemType))
}

func (itemType *ClipboardItemType) UnmarshalJSON(buf []byte) error {
	var itemTypeString string
	err := json.Unmarshal(buf, &itemTypeString)
	if err != nil {
		return err
	}

	switch itemTypeString {
	case "TEXT":
		*itemType = ClipboardItemTypeText
	case "IMAGE":
		*itemType = ClipboardItemTypeImage
	case "URL":
		*itemType = ClipboardItemTypeUrl
	case "UNKNOWN":
		*itemType = ClipboardItemTypeUnknown
	}

	return errors.New("unknown content item type '" + itemTypeString + "'")
}

type ClipboardItem struct {
	Id        string            `json:"id"`
	Type      ClipboardItemType `json:"type" ts_type:"'TEXT'|'IMAGE'|'URL'"`
	Content   string            `json:"content"`
	CreatedAt uint64            `json:"createdAt"`
	IsPinned  bool              `json:"isPinned"`
	PinnedAt  uint64            `json:"pinnedAt"`
}
